Add test case for init_rpc_backend

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26997 Add test case for init_rpc_backend**

- Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919
- Add test case to cover relevant lines.

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

xush6528 committed:
Fix init_rpc_backend callsite

Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

satgera commented on torch/distributed/rpc.py:
```diff
@@ -79,6 +79,7 @@ def _init_rpc(backend=RpcBackend.PROCESS_GROUP,
     elif is_rpc_backend_registered(backend):
         _agent = init_rpc_backend(
             backend,
+            self_rank=self_rank,
```
Can you add a test case where you register a backend and call init ?

xush6528 replied:
@satgera working on it. It requires me to patch init_rref_context

mrshenli reviewed (approved):
Would I be correct if I assume this is not needed for now as init_rpc_backend takes `*args` and `**args`, and the current registered backend we have does not really use `self_rank`? So this is only to make sure we do pass all args from `_init_rpc` to `init_rpc_backend`?

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

xush6528 commented:
@mrshenli open-source tests are still good.

I broke fb internal init_rpc_backend_handler. 
I will make the fb-internal init_rpc_backend_handler does not take self_rank in a following PR.

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

xush6528 commented:
disabled rpc test in py2.

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

xush6528 commented:
Checked. failed tests are test_sparse_allreduce_checks, test_cpp_cuda. not related to this PR.

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]

pietern commented on test/rpc_test.py:
```diff
@@ -1,24 +1,33 @@
 
-from torch.distributed.rpc import RpcBackend
 from common_utils import load_tests
+from torch.distributed.rpc_api import RpcBackend
```
I haven't seen this module before, is it new? If so, why was it added?

pietern replied:
Nevermind, I see it is added here. Why the rename?

xush6528 replied:
"rpc" module was shadowed by "rpc" method imported to "torch.distributed".
This caused the patch replacing the wrong target.
Fixed it by rename the module to "rpc_api".

pietern replied:
I see, thanks. I believe @rohan-varma is working on a PR to rename those to `rpc_sync` and `rpc_async`, right? Then this will no longer be necessary.

rohan-varma replied:
@pietern @xush6528 , that change is landed in https://github.com/pytorch/pytorch/pull/26570. But it doesn't remove the `rpc` function, it instead adds a warning to callers who use it and encourage them to call `rpc_sync` or `rpc_async`. We eventually want to entirely remove the `rpc` function as you mentioned, but wasn't sure if we could do this yet due to backwards compatibility issues, so after a discussion with @mrshenli (comments on the PR) we decided to go with this approach.

pietern replied:
@rohan-varma When we do, can you make sure to revert the rename? If it's an internal only module, it shouldn't be named `_api`, but rather use an underscore prefix, like all other private API like things in PyTorch.

rohan-varma replied:
@pietern, the `rpc_api.py` module is part of our external-facing API though, right? It contains things like `get_worker_id`, `remote, `rpc_sync`, etc. Should we define `__all__` to make it clear what our external facing APIs are?

xush6528 replied:
I feel ideally we want to put all external api names in `__init__.py` and has a module implementing these api names. 

It's a mess at the moment.
In `__init__.py` defines `def init_model_parallel`
In `rpc_api.py`, other API is defined (as Rohan listed).

rohan-varma replied:
@pietern @xush6528 , I've created #27234 to track this and explore possible ideas

xush6528 committed:
Update on "Fix init_rpc_backend callsite"


Reverting accidental change in https://github.com/pytorch/pytorch/pull/26919

Differential Revision: [D17637468](https://our.internmc.facebook.com/intern/diff/D17637468/)

[ghstack-poisoned]
